src/js_native_api_v8.cc

```diff
@@ -2256,6 +2259,65 @@ napi_status napi_create_external(napi_env env,
   return napi_clear_last_error(env);
 }
 
+NAPI_EXTERN napi_status napi_type_tag_object(napi_env env,
+                                             napi_value object,
+                                             const napi_type_tag* type_tag) {
+  NAPI_PREAMBLE(env);
+  v8::Local<v8::Context> context = env->context();
+  v8::Local<v8::Object> obj;
+  CHECK_TO_OBJECT_WITH_PREAMBLE(env, context, obj, object);
+
+  auto maybe_has = obj->HasPrivate(context,
+                                   NAPI_PRIVATE_KEY(context, type_tag));
+  CHECK_MAYBE_NOTHING_WITH_PREAMBLE(env, maybe_has, napi_generic_failure);
+  RETURN_STATUS_IF_FALSE_WITH_PREAMBLE(env,
+                                       !maybe_has.FromJust(),
+                                       napi_invalid_arg);
+
+  v8::Local<v8::ArrayBuffer> tag =
+      v8::ArrayBuffer::New(env->isolate, sizeof(*type_tag));
+  *static_cast<napi_type_tag*>(tag->GetContents().Data()) = *type_tag;
+
+  auto maybe_set = obj->SetPrivate(context,
+                                   NAPI_PRIVATE_KEY(context, type_tag),
+                                   tag);
+  CHECK_MAYBE_NOTHING_WITH_PREAMBLE(env, maybe_set, napi_generic_failure);
+  RETURN_STATUS_IF_FALSE_WITH_PREAMBLE(env,
+                                       maybe_set.FromJust(),
+                                       napi_generic_failure);
+
+  return GET_RETURN_STATUS(env);
+}
+
+NAPI_EXTERN napi_status
+napi_check_object_type_tag(napi_env env,
+                           napi_value object,
+                           const napi_type_tag* type_tag,
+                           bool* result) {
+  NAPI_PREAMBLE(env);
+  CHECK_ARG(env, result);
+  v8::Local<v8::Context> context = env->context();
+  v8::Local<v8::Object> obj;
+  CHECK_TO_OBJECT_WITH_PREAMBLE(env, context, obj, object);
+
+  auto maybe_value = obj->GetPrivate(context,
+                                     NAPI_PRIVATE_KEY(context, type_tag));
+  CHECK_MAYBE_EMPTY_WITH_PREAMBLE(env, maybe_value, napi_generic_failure);
+  v8::Local<v8::Value> js_type_tag = maybe_value.ToLocalChecked();
+
+  *result = false;
```

Comment:
Is there any reason not to make this assignment an `else` on the following `if` statement?

Reply:
I think my intention here is to express that we have passed the point where we might not even touch `*result` at all, because some checks could fail. I prefer that it be set unconditionally to false, because this is a runtime type check, and it should unconditionally fail by default.

I'm happy to change it to an else though, if you think this is not a good enough reason to keep it as is.

Reply:
I think my intention here is to express that we have passed the point where we might not even touch *result at all, because some checks could fail. I prefer that it be set unconditionally to false, because this is a runtime type check, and it should unconditionally fail by default.

I'm happy to change it to an else though, if you think this is not a good enough reason to keep it as is.